twine/utils.py

```diff
@@ -34,12 +39,18 @@
 # Shim for input to allow testing.
 input_func = input
 
-
 DEFAULT_REPOSITORY = "https://upload.pypi.org/legacy/"
 TEST_REPOSITORY = "https://test.pypi.org/legacy/"
 
+# TODO: In general, it seems to be assumed that the values retrieved from
+# instances of this type aren't None, except for username and password.
+# Type annotations would be cleaner if this were Dict[str, str], but that
+# requires reworking the username/password handling, probably starting with
+# get_userpass_value.
```

Comment:
> It might make sense to replace the value such that RepositoryConfig stores an auth = Optional[tuple(str, str)] instead of storing username and password separately. I suspect that's what session.auth expects.

That's correct

Reply:
If I follow, I *think* that would mean a type alias like:

```python
RepositoryConfig = Dict[str, Union[None, str, Tuple[str, str]]
```

which is starting to smell funny (unless require Python 3.8 and make `RepositoryConfig` a [`TypedDict`](https://www.python.org/dev/peps/pep-0589/) 😉).

I also suspect that implementing that could require more significant changes to the code, as opposed to the [current solution](https://github.com/pypa/twine/pull/519/files#diff-f5d1b38fc32763340f006b344c9a24d3R63-R70) of transforming the single values into the data structure that requests expects.

Reply:
Indeed, no need to tackle this now. Let's focus on typing what's already present and iterate from there.